### PR TITLE
Disable runfile tree creation

### PR DESCRIPTION
**Pull request by fzakaria**
Investigating why Bazel builds, specifically hydrating from Cache have gotten so bad seems to be on account of rules_python addition.

I debugged the cache step and found it caching duplication of python_rules: https://pastebin.com/ibgVB8Yk in the build tree.

These are in fact sym-links however:
```
❯ ls -l bazel-out/k8-fastbuild/bin/stablehlo/tests/vhlo/stablehlo_legalize_to_vhlo.0_9_0.mlir.test.runfiles/rules_python\~0.30.0\~python\~python_3_10_x86_64-unknown-linux-gnu/bin/
lrwxrwxrwx 175 780412 26 Feb 19:58 2to3 -> /usr/local/google/home/fmzakari/.cache/bazel/_bazel_fmzakari/17bce12c4b47a4a2fc75249afee05177/external/rules_python~0.30.0~python~python_3_10_x86_64-unknown-linux-gnu/bin/2to3
lrwxrwxrwx 180 780412 26 Feb 19:58 2to3-3.10 -> /usr/local/google/home/fmzakari/.cache/bazel/_bazel_fmzakari/17bce12c4b47a4a2fc75249afee05177/external/rules_python~0.30.0~python~python_3_10_x86_64-unknown-linux-gnu/bin/2to3-3.10
lrwxrwxrwx 176 780412 26 Feb 19:58 idle3 -> /usr/local/google/home/fmzakari/.cache/bazel/_bazel_fmzakari/17bce12c4b47a4a2fc75249afee05177/external/rules_python~0.30.0~python~python_3_10_x86_64-unknown-linux-gnu/bin/idle3
```

However these are a ton additional inodes that `tar` has to expand and store in the GitHub action.

People on the Bazel slack pointed me to "-nobuild_runfile_links" which disables the runfile **at build creation**.
Our tests however still run because `bazel test` has support for on-demand runfile generation.

This should make the caching much faster.

### CODE REVIEW

**Comment by mlevesquedion** [comment]
Change LGTM but this is making me wonder: do we even want caching? If we're going to pull this stuff in anyway and caching seems like it may be slower than rebuilding?

File: .bazelrc [hunk 1]
```diff
@@ -21,3 +21,10 @@ build --features=layering_check
 # Set the default compiler to the `clang` binary on the `PATH`.
 # TODO(fzakaria): Make this a toolchain or hermetic somehow
 build --repo_env=CC=clang
+
+# Disabling runfiles links drastically increases performance in slow disk IO situations
+# Do not build runfile trees by default. If an execution strategy relies on runfile
+# symlink teee, the tree is created on-demand. See: https://github.com/bazelbuild/bazel/issues/6627
+# and https://github.com/bazelbuild/bazel/commit/03246077f948f2790a83520e7dccc2625650e6df
+build --nobuild_runfile_links
+test --nobuild_runfile_links
```
Comment (mlevesquedion) [inline .bazelrc:30]: Nit: Looks like this file is missing a newline at EOF. The whitespace checks script only checks a subset of files, not sure why that is: https://github.com/openxla/stablehlo/blob/main/build_tools/github_actions/lint_whitespace_checks.sh#L43

Consider adding a newline.

**Review by mlevesquedion — APPROVED** [review]

File: .bazelrc [hunk 2]
```diff
@@ -21,3 +21,10 @@ build --features=layering_check
 # Set the default compiler to the `clang` binary on the `PATH`.
 # TODO(fzakaria): Make this a toolchain or hermetic somehow
 build --repo_env=CC=clang
+
+# Disabling runfiles links drastically increases performance in slow disk IO situations
+# Do not build runfile trees by default. If an execution strategy relies on runfile
+# symlink teee, the tree is created on-demand. See: https://github.com/bazelbuild/bazel/issues/6627
```
Comment (mlevesquedion) [inline .bazelrc:27]:
```suggestion
# symlink tree, the tree is created on-demand. See: https://github.com/bazelbuild/bazel/issues/6627
```

**Comment by fzakaria** [comment]
The cache also includes whether we passed a test which cuts down on rebuilding the code.

File: stablehlo/testdata/BUILD.bazel [hunk 3]
```diff
@@ -28,7 +28,7 @@ expand_template(
         "\"@STABLEHLO_SOURCE_DIR@\"": "os.path.join(os.environ['TEST_SRCDIR'], '_main')",
         "\"@STABLEHLO_TOOLS_DIR@\"": "os.path.join(os.environ['TEST_SRCDIR'], '_main')",
         "@LIT_SITE_CFG_IN_HEADER@": "# Autogenerated, do not edit.",
-        "@LLVM_TOOLS_DIR@": package_path("@llvm-project//llvm:BUILD"),
+        "@LLVM_TOOLS_DIR@": "../llvm-project/llvm",
```
Comment (fzakaria) [inline stablehlo/testdata/BUILD.bazel:31]: I filed https://github.com/llvm/llvm-project/issues/83066 -- this cuts down on the # of symlinks effectively by half.

**Pull request merged**